### PR TITLE
stats: use leased types in the stats cache

### PR DESCRIPTION
**Pull request by rohany**
This commit switches the stats cache to use leased types rather than
accessing types directly. This was the final user of the direct type
access methods, so that code is removed as well.

Release note: None

### CODE REVIEW

**Review by ajwerner — CHANGES_REQUESTED** [review]
> Reviewed 5 of 7 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53078)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @RaduBerinde, and @rohany)

___
*[pkg/sql/stats/stats_cache.go, line 419 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53078#-MF7dw9OEX6kRQKzdKhr:-MF7dw9OEX6kRQKzdKhs:b-wzom3d) ([raw file](https://github.com/cockroachdb/cockroach/blob/45f9fd348c314a91effa73cadda9dd495877015a/pkg/sql/stats/stats_cache.go#L419)):*
> ```Go
> 			// collecting the stats. Changes to types are backwards compatible across
> 			// versions, so using a newer version of the type metadata here is safe.
> 			// Additionally, the leased type that is accessed here is guaranteed to
> ```

This comment and code seems pretty suspect. You're guaranteed to be able to read the values of the type while you hold the lease and the transaction's read timestamp doesn't get pushed above the lease's deadline. the fact that this transaction isn't tied to the lifetime of the request which read the data seems odd. One thing that would work would be to pass in an hlc timestamp that was used to create the stats (I'm pretty sure they are created at a timestamp), and then to set the transaction timestamp here to that fixed timestamp. Then I think this mostly works. The last remaining problem is that it's not clear to me that we're releasing the reference to the lease that we acquire underneath this new collection.

**Comment by rohany** [comment]
I'm not too keen on even using the lease manager here honestly, I just wanted to get rid of the last user of `ResolveTypeDescByID`. I might just update this to get the type descriptor directly using a txn.

**Review by ajwerner — COMMENTED** [review]
I think the lease manager is ... fine. I'd just like to see a reasonably timestamp make its way here if we're going to make claims. I feel like given the fact that you can't delete an enum value, I'd be okay if you just say that so long as you find any descriptor, you'll be able to deal with these stats and thus it doesn't matter if it isn't the exact enum descriptor used to collect the stats. 

tl;dr if you fix the comment to explain things reasonably I'm on board. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53078#-:-MFB7TLI1ZEHXqSgHfR4:b-qxjid6)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde and @rohany)

**Review by ajwerner — APPROVED** [review]
:lgtm: 

> Reviewed 1 of 7 files at r1, 3 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53078#-:-MFCAUOzEe0EymJ2xxPy:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @RaduBerinde)
